File: 백준/Gold/14307. Monster Path （Large）/Monster Path （Large）.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;
using ld = long double;
const int dy[] = {-1, 0, 1, 0}, dx[] = {0, 1, 0, -1};

char board[25][25];
ld d[25][25];

ld solved(){
    int r, c, sy, sx, s; cin >> r >> c >> sy >> sx >> s; sy++, sx++;
    ld p, q; cin >> p >> q;
    for(int i = 1; i <= r; i++) for(int j = 1; j <= c; j++){
        cin >> board[i][j];
        if(board[i][j] == 'A') d[i][j] = p;
        else d[i][j] = q; 
    }
    ld ans = 0;
    for(int l = 0; l < (1<<(2*s)); l++){
        int y = sy, x = sx; ld sump = 0; bool flag = 1;
        for(int i = 0; i < s; i++){
            int k = (l>>(i*2))&3;
            y += dy[k]; x += dx[k];
            if(y < 1 || y > r || x < 1 || x > c){
                flag = 0; continue;
            }
            sump += d[y][x];
            if(board[y][x] == 'A') d[y][x] *= 1-p;
            else d[y][x] *= 1-q;
        }
        if(flag) ans = max(ans, sump);
        y = sy, x = sx;
        for(int i = 0; i < s; i++){
            int k = (l>>(i*2))&3;
            y += dy[k]; x += dx[k];
            if(y < 1 || y > r || x < 1 || x > c) continue;
            sump += d[y][x];
            if(board[y][x] == 'A') d[y][x] = p;
            else d[y][x] = q;
        }
    }
    return ans;
}

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int T; cin >> T; cout << fixed << setprecision(12);
    for(int i = 1; i <= T; i++) cout << "Case #" << i << ": " << solved() << '\n';
    return 0;
}

File: 백준/Gold/14307. Monster Path （Large）/Monster Path （Large）.py
import sys
def print(*args, sep=" ", end="\n"): return sys.stdout.write(sep.join(map(str, args)) + end)
def input(): return sys.stdin.readline().rstrip()

def main():
    T = int(input())
    def sv():
        r, c, sy, sx, s = map(int, input().split()); sy += 1; sx += 1
        p, q = map(float, input().split())
        board = [['']*(c+1) for _ in range(r+1)]
        d = [[0.0]*(c+1) for _ in range(r+1)]
        dy = (-1, 0, 1, 0); dx = (0, 1, 0, -1)
        for i in range(1, r+1):
            row = input().split()
            for j in range(1, c+1):
                board[i][j] = row[j-1]
                if board[i][j] == 'A': d[i][j] = p
                else: d[i][j] = q
        ans = 0
        for l in range(1<<(2*s)):
            y = sy; x = sx; sump = 0; flag = 1
            for i in range(s):
                k = (l>>(i*2))&3
                y += dy[k]; x += dx[k]
                if y < 1 or  y > r or x < 1 or x > c: flag = 0; continue
                sump += d[y][x]
                if board[y][x] == 'A': d[y][x] *= 1-p
                else: d[y][x] *= 1-q
            if flag: ans = max(ans, sump)
            y = sy; x = sx
            for i in range(s):
                k = (l>>(i*2))&3
                y += dy[k]; x += dx[k]
                if y < 1 or  y > r or x < 1 or x > c: continue
                if board[y][x] == 'A': d[y][x] = p
                else: d[y][x] = q
        return ans
    for i in range(1, T+1): print(f'Case #{i}: {sv():.7f}')
    return 0
if __name__ == '__main__':
    sys.exit(main())